// magister.js/lib/activity.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _magisterThing = require('./magisterThing');

var _magisterThing2 = _interopRequireDefault(_magisterThing);

var _activityElement = require('./activityElement');

var _activityElement2 = _interopRequireDefault(_activityElement);

var _util = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @extends {MagisterThing}
 */
class Activity extends _magisterThing2.default {
	/**
  * @private
  * @param {Magister} magister
  * @param {Object} raw
  */
	constructor(magister, raw) {
		super(magister);

		/**
   * @type {String}
   * @readonly
   */
		this.id = (0, _util.toString)(raw.Id);
		/**
   * @type {String}
   * @readonly
   */
		this.title = raw.Titel;
		/**
   * @type {String}
   * @readonly
   */
		this.description = (0, _util.cleanHtmlContent)(raw.Details);

		/**
   * @type {Date}
   * @readonly
   */
		this.signinStart = (0, _util.parseDate)(raw.StartInschrijfdatum);
		/**
   * @type {Date}
   * @readonly
   */
		this.signinEnd = (0, _util.parseDate)(raw.EindeInschrijfdatum);

		/**
   * @type {Date}
   * @readonly
   */
		this.visibleFrom = (0, _util.parseDate)(raw.ZichtbaarVanaf);
		/**
   * @type {Date}
   * @readonly
   */
		this.visibleTo = (0, _util.parseDate)(raw.ZichtbaarTotEnMet);

		/**
   * @type {String}
   * @private
   * @readonly
   */
		this._url = `${magister._personUrl}/activiteiten/${this.id}`;
	}

	/**
  * @return {Promise<ActivityElement[]>}
  */
	elements() {
		const url = `${this._url}/onderdelen`;
		return this._magister.http.get(url).then(res => res.json()).then(res => res.Items.map(e => new _activityElement2.default(this._magister, this, e)));
	}
}

exports.default = Activity;
