// magister.js/lib/addressInfo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _magisterThing = _interopRequireDefault(require("./magisterThing"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class AddressInfo extends _magisterThing.default {
  /**
   * @private
   * @param {Magister} magister
   * @param {Object} raw
   */
  constructor(magister, raw) {
    super(magister);
    /**
     * @type {String}
     * @readonly
     */

    this.postalCode = raw.Postcode;
    /**
     * @type {String}
     * @readonly
     */

    this.street = raw.Straatnaam;
    /**
     * @type {Number}
     * @readonly
     */

    this.houseNumber = raw.Huisnummer;
    /**
     * String behind the `houseNumber` (eg 'A')
     * @type {String}
     * @readonly
     */

    this.suffix = raw.Toevoeging;
    /**
     * @type {String}
     * @readonly
     */

    this.city = raw.Woonplaats;
  }

}

var _default = AddressInfo;
exports.default = _default;
