// magister.js/lib/gradePeriod.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _magisterThing = _interopRequireDefault(require("./magisterThing"));

var _util = require("./util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class GradePeriod extends _magisterThing.default {
  /**
   * @private
   * @param {Magister} magister
   * @param {Object} raw
   */
  constructor(magister, raw) {
    super(magister);
    /**
     * @type {String}
     * @readonly
     */

    this.id = (0, _util.toString)(raw.Id);
    /**
     * @type {Date}
     * @readonly
     */

    this.begin = (0, _util.parseDate)(raw.Start);
    /**
     * @type {Date}
     * @readonly
     */

    this.end = (0, _util.parseDate)(raw.Einde);
    /**
     * @type {String}
     * @readonly
     * @default ''
     */

    this.name = _lodash.default.trim(raw.Naam);
    /**
     * @type {String}
     * @readonly
     * @default ''
     */

    this.description = _lodash.default.trim(raw.Omschrijving);
  }

}

var _default = GradePeriod;
exports.default = _default;
