// magister.js/lib/schoolUtility.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _magisterThing = _interopRequireDefault(require("./magisterThing"));

var _class = _interopRequireDefault(require("./class"));

var _util = require("./util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class SchoolUtility extends _magisterThing.default {
  /**
   * @private
   * @param {Magister} magister
   * @param {Object} raw
   */
  constructor(magister, raw) {
    super(magister);
    /**
     * @type {String}
     * @readonly
     */

    this.id = toString(raw.Id);
    /**
     * @type {Number}
     * @readonly
     */

    this.type = raw.MateriaalType; // REVIEW: do we want a string getter for this?

    /**
     * @type {String}
     * @readonly
     */

    this.name = raw.Titel;
    /**
     * @type {String}
     * @readonly
     */

    this.publisher = raw.Uitgeverij;
    /**
     * @type {Number}
     * @readonly
     */

    this.state = raw.Status;
    /**
     * @type {Date}
     * @readonly
     */

    this.begin = (0, _util.parseDate)(raw.Start);
    /**
     * @type {Date}
     * @readonly
     */

    this.end = (0, _util.parseDate)(raw.Eind);
    /**
     * @type {String}
     * @readonly
     */

    this.EAN = raw.EAN;
    /**
     * @type {String}
     * @readonly
     */

    this.url = function () {
      const link = raw.Links.find(l => l.Rel === 'content');
      return link == null ? undefined : link.Href;
    }();
    /**
     * @type {Class}
     * @readonly
     */


    this.class = new _class.default(magister, raw.Vak);
  }

}

var _default = SchoolUtility;
exports.default = _default;
